src/auxiliary/auxiliary.jl

```diff
@@ -132,21 +132,72 @@ get_name(::Val{x}) where x = string(x)
 """
     @threaded for ... end
 
-Basically the same as `Threads.@threads` with an additional check whether only
-one thread is used to reduce the overhead of serial execution.
+Semantically the same as `Threads.@threads` when iterating over a `AbstractUnitRange`
+but without guarantee that the underlying implementation uses `Threads.@threads`
+or works for more general for loops.
+In particular, there may be an additional check whether only one thread is used
+to reduce the overhead of serial execution or the underlying threading capabilities
+might be provided by other packages such as [CheapThreads.jl](https://github.com/JuliaSIMD/CheapThreads.jl).
+
+!!! warn
+  This macro does not necessarily work for general `for` loops.
 
 Some discussion can be found at https://discourse.julialang.org/t/overhead-of-threads-threads/53964
+and https://discourse.julialang.org/t/threads-threads-with-one-thread-how-to-remove-the-overhead/58435.
 """
 macro threaded(expr)
-  # esc(quote ... end) as suggested in https://github.com/JuliaLang/julia/issues/23221
-  return esc(quote
-    let
-      if Threads.nthreads() == 1
-        $(expr)
-      else
-        Threads.@threads $(expr)
+  # Use `esc(quote ... end)` for nested macro calls as suggested in
+  # https://github.com/JuliaLang/julia/issues/23221
+  #
+  # The following code is a simple version using only `Threads.@threads` from the
+  # standard library with an additional check whether only a single thread is used
+  # to reduce some overhead (and allocations) for serial execution.
+  #
+  # return esc(quote
+  #   let
+  #     if Threads.nthreads() == 1
+  #       $(expr)
+  #     else
+  #       Threads.@threads $(expr)
+  #     end
+  #   end
+  # end)
+  #
+  # However, the code below using `@batch` from CheapThreads.jl is more efficient,
+  # since this packages provides threads with less overhead. Since it is written
+  # by Chris Elrod, the author of LoopVectorization.jl, we expect this package
+  # to provide the most efficient and useful implementation of threads (as we use
+  # them) available in Julia.
+  # !!! danger "Heisenbug"
+  #     Look at the comments for `wrap_array` when considering to change this macro.
+
+  return esc(quote @batch $(expr) end)
```

Comment:
I'm sorry I don't understand exactly what you mean, @chriselrod. Are you referring to the additional check whether only one thread is used mentioned in the docstring? We like it and make use of it by just forwarding `@threaded` to `@batch`. The other implementation using such a check and `Threads.@threads` which is commented out should basically serve as a demonstration of what kind of optimizations might be happening.
The basic reason why we do not use `@batch` directly everywhere in our source code is that we had already introduced `@threaded` before so that it's part of the documented API right now.

Reply:
FWIW, the check `(CheapThreads.CheapThreads).num_threads() == 1` should be 0-overhead because `num_threads()` gets redefined when the number of threads changes:
```julia
julia> using VectorizationBase

julia> Threads.nthreads()
1

julia> VectorizationBase.num_threads()
static(1)

julia> @code_typed VectorizationBase.num_threads()
CodeInfo(
1 ─     return static(1)
) => StaticInt{1}
```
```julia
julia> using VectorizationBase

julia> Threads.nthreads()
2

julia> VectorizationBase.num_threads()
static(2)

julia> @code_typed VectorizationBase.num_threads()
CodeInfo(
1 ─     return static(2)
) => StaticInt{2}

julia> @code_typed Threads.nthreads()
CodeInfo(
1 ─ %1 = Base.Threads.cglobal(:jl_n_threads, Base.Threads.Cint)::Ptr{Int32}
│   %2 = Base.pointerref(%1, 1, 1)::Int32
│   %3 = Core.sext_int(Core.Int64, %2)::Int64
└──      return %3
) => Int64
```

Reply:
That's really good to see :+1: